File: src/lib/resolver.ts
import { normalize, looksLikeUrl } from "./phonetic";

const KNOWN_DOMAINS = [
  "google.com", "youtube.com", "facebook.com", "twitter.com", "instagram.com",
  "linkedin.com", "reddit.com", "github.com", "stackoverflow.com", "wikipedia.org",
  "amazon.com", "netflix.com", "spotify.com", "apple.com", "microsoft.com",
  "onet.pl", "wp.pl", "interia.pl", "gazeta.pl", "allegro.pl",
  "olx.pl", "tvn24.pl", "bankier.pl", "money.pl", "pudelek.pl",
  "wykop.pl", "naszemiasto.pl", "o2.pl", "tvp.pl", "polsat.pl",
  "ceneo.pl", "morele.net", "x-kom.pl", "mediaexpert.pl", "empik.com",
  "bbc.com", "cnn.com", "nytimes.com", "theguardian.com", "reuters.com",
  "yahoo.com", "bing.com", "duckduckgo.com", "twitch.tv", "tiktok.com",
  "pinterest.com", "tumblr.com", "discord.com", "slack.com", "zoom.us",
  "medium.com", "dev.to", "hackernews.com", "producthunt.com",
  "gitlab.com", "bitbucket.org", "npmjs.com", "pypi.org",
  "aws.amazon.com", "azure.microsoft.com", "cloud.google.com",
  "docs.google.com", "drive.google.com", "maps.google.com",
  "mail.google.com", "calendar.google.com", "translate.google.com",
  "ebay.com", "aliexpress.com", "wish.com", "etsy.com",
  "booking.com", "airbnb.com", "tripadvisor.com",
  "imdb.com", "rottentomatoes.com", "metacritic.com",
  "wordpress.com", "blogger.com", "wix.com", "squarespace.com",
  "dropbox.com", "onedrive.com", "mega.nz",
  "paypal.com", "stripe.com", "revolut.com",
  "openai.com", "anthropic.com", "huggingface.co",
];

export interface ResolveResult {
  url: string | null;
  suggestions: string[];
  resolveType: "exact" | "fuzzy" | "search" | "ambiguous";
  needsClarification: boolean;
  normalizedInput: string;
}

function similarity(a: string, b: string): number {
  if (a === b) return 1;
  const longer = a.length > b.length ? a : b;
  const shorter = a.length > b.length ? b : a;
  if (longer.length === 0) return 1;

  const costs: number[] = [];
  for (let i = 0; i <= longer.length; i++) {
    let lastValue = i;
    for (let j = 0; j <= shorter.length; j++) {
      if (i === 0) {
        costs[j] = j;
      } else if (j > 0) {
        let newValue = costs[j - 1];
        if (longer[i - 1] !== shorter[j - 1]) {
          newValue = Math.min(Math.min(newValue, lastValue), costs[j]) + 1;
        }
        costs[j - 1] = lastValue;
        lastValue = newValue;
      }
    }
    if (i > 0) costs[shorter.length] = lastValue;
  }
  return (longer.length - costs[shorter.length]) / longer.length;
}

function fuzzyMatchDomain(
  input: string,
  threshold = 0.55,
  maxResults = 6,
): Array<[string, number]> {
  let cleaned = input.toLowerCase().trim();
  cleaned = cleaned.replace(/^(https?:\/\/|www\.)/, "");
  cleaned = cleaned.split("/")[0];

  const matches: Array<[string, number]> = [];
  for (const domain of KNOWN_DOMAINS) {
    const ratio = similarity(cleaned, domain);
    if (ratio >= threshold) {
      matches.push([domain, ratio]);
    }
  }

  matches.sort((a, b) => b[1] - a[1]);
  return matches.slice(0, maxResults);
}

export function resolve(rawInput: string, threshold = 0.55): ResolveResult {
  const text = rawInput.trim();
  if (!text) {
    return {
      url: null,
      suggestions: [],
      resolveType: "ambiguous",
      needsClarification: true,
      normalizedInput: "",
    };
  }

  // 1) Already a valid URL?
  if (/^https?:\/\//.test(text)) {
    return {
      url: text,
      suggestions: [],
      resolveType: "exact",
      needsClarification: false,
      normalizedInput: text,
    };
  }

  // 2) Looks like a domain?
  if (/^[\w.-]+\.\w{2,}/.test(text)) {
    return {
      url: `https://${text}`,
      suggestions: [],
      resolveType: "exact",
      needsClarification: false,
      normalizedInput: text,
    };
  }

  // 3) Apply phonetic normalization
  const normalized = normalize(text);
  if (looksLikeUrl(normalized)) {
    const fuzzy = fuzzyMatchDomain(normalized, threshold);
    const suggestions = fuzzy
      .filter(([d]) => d !== normalized)
      .slice(0, 3)
      .map(([d]) => `https://${d}`);
    return {
      url: `https://${normalized}`,
      suggestions,
      resolveType: "fuzzy",
      needsClarification: false,
      normalizedInput: normalized,
    };
  }

  // 4) Fuzzy match against known domains
  const fuzzy = fuzzyMatchDomain(text, threshold);
  if (fuzzy.length > 0) {
    const [, bestScore] = fuzzy[0];
    const allSuggestions = fuzzy.map(([d]) => `https://${d}`);

    if (bestScore > 0.8) {
      return {
        url: allSuggestions[0],
        suggestions: allSuggestions.slice(1, 4),
        resolveType: "fuzzy",
        needsClarification: false,
        normalizedInput: text,
      };
    } else {
      return {
        url: null,
        suggestions: allSuggestions.slice(0, 5),
        resolveType: "ambiguous",
        needsClarification: true,
        normalizedInput: text,
      };
    }
  }

  // 5) Fallback — search
  const searchUrl = `https://duckduckgo.com/?q=${encodeURIComponent(text)}`;
  return {
    url: searchUrl,
    suggestions: [],
    resolveType: "search",
    needsClarification: false,
    normalizedInput: text,
  };
}

File: resolver.py
"""
URL Resolver — converts raw user input (text or speech) into valid URLs.

Pipeline:
  1. Direct URL detection (https://...)
  2. Domain detection (domain.tld)
  3. Phonetic normalization → domain detection
  4. Fuzzy matching against known domains
  5. Fallback to search engine query
"""

import re
from dataclasses import dataclass, field
from difflib import SequenceMatcher
from urllib.parse import quote_plus

from .phonetic import normalize, looks_like_url
from .domains import KNOWN_DOMAINS


@dataclass
class ResolveResult:
    """Result of URL resolution."""
    url: str | None
    suggestions: list[str] = field(default_factory=list)
    resolve_type: str = "exact"  # exact | fuzzy | search | ambiguous
    needs_clarification: bool = False
    normalized_input: str = ""


def fuzzy_match_domain(
    input_str: str,
    threshold: float = 0.55,
    max_results: int = 6,
) -> list[tuple[str, float]]:
    """
    Find closest matching domains using SequenceMatcher.

    Args:
        input_str: Raw or normalized domain string.
        threshold: Minimum similarity score (0.0–1.0).
        max_results: Maximum number of results.

    Returns:
        List of (domain, score) tuples, sorted by score descending.
    """
    cleaned = input_str.lower().strip()
    cleaned = re.sub(r"^(https?://|www\.)", "", cleaned)
    cleaned = cleaned.split("/")[0]

    matches = []
    for domain in KNOWN_DOMAINS:
        ratio = SequenceMatcher(None, cleaned, domain).ratio()
        if ratio >= threshold:
            matches.append((domain, ratio))

    matches.sort(key=lambda x: -x[1])
    return matches[:max_results]


def resolve(raw_input: str, threshold: float = 0.55) -> ResolveResult:
    """
    Resolve user input into a browseable URL.

    Args:
        raw_input: Raw text from keyboard or speech recognition.
        threshold: Fuzzy matching threshold.

    Returns:
        ResolveResult with URL, suggestions, and resolution type.
    """
    text = raw_input.strip()
    if not text:
        return ResolveResult(url=None, resolve_type="ambiguous", needs_clarification=True)

    # 1) Already a valid URL?
    if re.match(r"^https?://", text):
        return ResolveResult(url=text, resolve_type="exact", normalized_input=text)

    # 2) Looks like a domain?
    if re.match(r"^[\w.-]+\.\w{2,}", text):
        return ResolveResult(
            url=f"https://{text}",
            resolve_type="exact",
            normalized_input=text,
        )

    # 3) Apply phonetic normalization
    normalized = normalize(text)
    if looks_like_url(normalized):
        # Check fuzzy matches too, in case normalization was imperfect
        fuzzy = fuzzy_match_domain(normalized, threshold)
        suggestions = [f"https://{d}" for d, _ in fuzzy if d != normalized]
        return ResolveResult(
            url=f"https://{normalized}",
            suggestions=suggestions[:3],
            resolve_type="fuzzy",
            normalized_input=normalized,
        )

    # 4) Fuzzy match against known domains
    fuzzy = fuzzy_match_domain(text, threshold)
    if fuzzy:
        best_domain, best_score = fuzzy[0]
        all_suggestions = [f"https://{d}" for d, _ in fuzzy]

        if best_score > 0.80:
            return ResolveResult(
                url=all_suggestions[0],
                suggestions=all_suggestions[1:4],
                resolve_type="fuzzy",
                normalized_input=text,
            )
        else:
            return ResolveResult(
                url=None,
                suggestions=all_suggestions[:5],
                resolve_type="ambiguous",
                needs_clarification=True,
                normalized_input=text,
            )

    # 5) Fallback — treat as search query
    search_url = f"https://duckduckgo.com/?q={quote_plus(text)}"
    return ResolveResult(
        url=search_url,
        resolve_type="search",
        normalized_input=text,
    )
